presto-main/src/main/java/io/prestosql/sql/planner/iterative/rule/UnwrapCastInComparison.java

```diff
@@ -0,0 +1,397 @@
+ *
+ * <pre>
+ * x = smallint '1'
+ * </pre>
+ *
+ * It can simplify expressions that are known to be true or false, and
+ * remove the comparisons altogether. For example, give x::smallint,
+ * for an expression like:
+ *
+ * <pre>
+ * CAST(x AS bigint) > bigint '10000000'
+ *</pre>
+ */
+public class UnwrapCastInComparison
+        extends ExpressionRewriteRuleSet
+{
+    public UnwrapCastInComparison(Metadata metadata, TypeAnalyzer typeAnalyzer)
+    {
+        super(createRewrite(metadata, typeAnalyzer));
+    }
+
+    private static ExpressionRewriter createRewrite(Metadata metadata, TypeAnalyzer typeAnalyzer)
+    {
+        requireNonNull(metadata, "metadata is null");
+        requireNonNull(typeAnalyzer, "typeAnalyzer is null");
+
+        return (expression, context) -> {
+            if (SystemSessionProperties.isUnwrapCastOptimizationEnabled(context.getSession())) {
+                return ExpressionTreeRewriter.rewriteWith(new Visitor(metadata, typeAnalyzer, context.getSession(), context.getSymbolAllocator().getTypes()), expression);
+            }
+
+            return expression;
+        };
+    }
+
+    private static class Visitor
+            extends io.prestosql.sql.tree.ExpressionRewriter<Void>
+    {
+        private final Metadata metadata;
+        private final TypeAnalyzer typeAnalyzer;
+        private final Session session;
+        private final TypeProvider types;
+        private final InterpretedFunctionInvoker functionInvoker;
+        private final LiteralEncoder literalEncoder;
+
+        public Visitor(Metadata metadata, TypeAnalyzer typeAnalyzer, Session session, TypeProvider types)
+        {
+            this.metadata = requireNonNull(metadata, "metadata is null");
+            this.typeAnalyzer = requireNonNull(typeAnalyzer, "typeAnalyzer is null");
+            this.session = requireNonNull(session, "session is null");
+            this.types = requireNonNull(types, "types is null");
+            this.functionInvoker = new InterpretedFunctionInvoker(metadata.getFunctionRegistry());
+            this.literalEncoder = new LiteralEncoder(metadata.getBlockEncodingSerde());
+        }
+
+        @Override
+        public Expression rewriteComparisonExpression(ComparisonExpression node, Void context, ExpressionTreeRewriter<Void> treeRewriter)
+        {
+            ComparisonExpression expression = (ComparisonExpression) treeRewriter.defaultRewrite((Expression) node, null);
+            return unwrapCast(expression);
+        }
+
+        private Expression unwrapCast(ComparisonExpression expression)
+        {
+            if (!(expression.getLeft() instanceof Cast)) {
+                return expression;
+            }
+
+            Object right = ExpressionInterpreter.expressionOptimizer(expression.getRight(), metadata, session, typeAnalyzer.getTypes(session, types, expression.getRight()))
+                    .optimize(NoOpSymbolResolver.INSTANCE);
+
+            if (right instanceof Expression) {
```

Comment:
can `.optimize()` return `NullExpression` instead of `null`?

Reply:
No, but I'll add a check to avoid future breakage in case something changes.